File: backend/app/schemas/commission.py
from decimal import Decimal
from datetime import date, datetime
from typing import Optional, List
from pydantic import BaseModel, Field


class CampaignTypeBase(BaseModel):
    category: str
    campaign_name: str = Field(..., max_length=255)
    description: Optional[str] = None
    is_active: bool = True
    valid_from: Optional[date] = None
    valid_to: Optional[date] = None

class CampaignTypeCreate(CampaignTypeBase):
    pass

class CampaignTypeSchema(CampaignTypeBase):
    id: int
    created_at: datetime
    transaction_count: Optional[int] = None

    class Config:
        from_attributes = True


class StatementBatchBase(BaseModel):
    statement_date: date
    house_id: int
    batch_reference: str = Field(..., max_length=100)

class StatementBatchCreate(StatementBatchBase):
    pass

class StatementBatchSchema(StatementBatchBase):
    id: int
    total_records: Optional[int] = 0
    processed_records: Optional[int] = 0
    failed_records: Optional[int] = 0
    status: str
    error_log: Optional[str] = None
    uploaded_by: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class CampaignTransactionBase(BaseModel):
    campaign_type_id: int
    participant_type: str
    participant_ref: str = Field(..., max_length=100)
    participant_name: Optional[str] = None
    purpose: Optional[str] = None
    amount: Decimal = Field(..., ge=0)
    extra_data: Optional[dict] = {}

class CampaignTransactionCreate(CampaignTransactionBase):
    pass

class CampaignTransactionUpdate(BaseModel):
    campaign_type_id: Optional[int] = None
    participant_type: Optional[str] = None
    participant_ref: Optional[str] = Field(None, max_length=100)
    participant_name: Optional[str] = None
    purpose: Optional[str] = None
    amount: Optional[Decimal] = Field(None, ge=0)
    extra_data: Optional[dict] = None

class CampaignTransactionSchema(CampaignTransactionBase):
    id: int
    statement_batch_id: int
    house_id: int
    created_at: datetime
    campaign_type: Optional[CampaignTypeSchema] = None

    class Config:
        from_attributes = True


class DateFilter(BaseModel):
    exact: Optional[date] = None
    from_date: Optional[date] = Field(None, alias="from")
    to_date: Optional[date] = Field(None, alias="to")
    month: Optional[int] = Field(None, ge=1, le=12)
    year: Optional[int] = None

class CommissionFilterRequest(BaseModel):
    date: Optional[DateFilter] = None
    house_ids: Optional[List[int]] = None
    campaign_type_ids: Optional[List[int]] = None
    campaign_category: Optional[str] = None
    participant_type: Optional[str] = None
    search: Optional[str] = Field(None, max_length=200)
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=50, ge=1, le=500)
    sort_by: Optional[str] = "created_at"
    sort_order: Optional[str] = "desc"


class CommissionSummary(BaseModel):
    total_campaign_amount: Decimal = Decimal("0.00")
    transaction_count: int = 0
    house_count: int = 0


class CampaignPerformance(BaseModel):
    campaign_type_id: int
    campaign_name: str
    category: str
    total_amount: Decimal
    transaction_count: int
    house_count: int


class HousePerformance(BaseModel):
    house_id: int
    house_code: str
    house_name: str
    total_amount: Decimal
    transaction_count: int


class PaginatedResponse(BaseModel):
    items: List
    total: int
    page: int
    page_size: int
    total_pages: int
    summary: Optional[CommissionSummary] = None


class CommissionImportResponse(BaseModel):
    batch_reference: str
    total_rows: int
    valid_rows: int
    failed_rows: int
    errors: Optional[List[dict]] = None


class DashboardAnalytics(BaseModel):
    summary: CommissionSummary
    campaign_performance: List[CampaignPerformance]
    house_performance: List[HousePerformance]
    monthly_trend: Optional[List[dict]] = None

File: frontend/src/types/commission.ts
export interface CommissionFilterState {
  date: {
    from: string;
    to: string;
  };
  houseIds: number[];
  campaignTypeIds: number[];
  campaignCategory: string;
  participantType: string;
  search: string;
}

export interface DateFilter {
  exact?: string;
  from?: string;
  to?: string;
  month?: number;
  year?: number;
}

export interface CommissionFilterPayload {
  date?: DateFilter;
  house_ids?: number[];
  campaign_type_ids?: number[];
  campaign_category?: string;
  participant_type?: string;
  search?: string;
  page: number;
  page_size: number;
  sort_by?: string;
  sort_order?: string;
}

export interface CommissionTransaction {
  id: number;
  statement_date: string;
  batch_reference: string;
  house_id: number;
  house_code: string;
  house_name: string;
  campaign_name: string;
  campaign_category: string;
  participant_type: string;
  participant_ref: string;
  participant_name: string;
  employee_id: number | null;
  employee_employee_id: string | null;
  employee_dms_code: string | null;
  employee_name: string | null;
  purpose: string | null;
  amount: number;
  extra_data: Record<string, unknown> | null;
  created_at: string;
}

export interface CommissionSummary {
  total_campaign_amount: number;
  transaction_count: number;
  house_count: number;
}

export interface CampaignPerformance {
  campaign_type_id: number;
  campaign_name: string;
  category: string;
  total_amount: number;
  transaction_count: number;
  house_count: number;
}

export interface HousePerformance {
  house_id: number;
  house_code: string;
  house_name: string;
  total_amount: number;
  transaction_count: number;
}

export interface PaginatedResponse {
  items: CommissionTransaction[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  summary: CommissionSummary;
}

export interface DashboardAnalytics {
  summary: CommissionSummary;
  campaign_performance: CampaignPerformance[];
  house_performance: HousePerformance[];
  monthly_trend?: MonthlyTrend[];
}

export interface MonthlyTrend {
  year: number;
  month: number;
  campaign_total: string;
}

export interface FilterOption {
  id: number;
  code?: string;
  name?: string;
}

export interface FilterOptions {
  houses: FilterOption[];
  campaigns: { id: number; name: string; category: string }[];
  available_months: string[];
  categories: string[];
  participant_types: string[];
}

export interface ImportResponse {
  batch_reference: string;
  total_rows: number;
  valid_rows: number;
  failed_rows: number;
  errors?: { row: number; errors: string[] }[];
}

export interface CommissionTransactionUpdate {
  campaign_type_id?: number;
  participant_type?: string;
  participant_ref?: string;
  participant_name?: string;
  purpose?: string;
  amount?: number;
  extra_data?: Record<string, unknown>;
}

export const DEFAULT_FILTER_STATE: CommissionFilterState = {
  date: { from: "", to: "" },
  houseIds: [],
  campaignTypeIds: [],
  campaignCategory: "",
  participantType: "",
  search: "",
};
